src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs

```diff
@@ -223,48 +261,13 @@ protected virtual BoundStatement GenerateSetResultCall()
                         : ImmutableArray<BoundExpression>.Empty));
         }
 
-        protected BoundCatchBlock GenerateExceptionHandling(LocalSymbol exceptionLocal, ImmutableArray<StateMachineFieldSymbol> hoistedLocals)
-        {
-            // catch (Exception ex)
-            // {
-            //     _state = finishedState;
-            //
-            //     for each hoisted local:
-            //     <>x__y = default
-            //
-            //     builder.SetException(ex);  OR  if (this.combinedTokens != null) this.combinedTokens.Dispose(); _promiseOfValueOrEnd.SetException(ex); /* for async-iterator method */
-            //     return;
-            // }
-
-            // _state = finishedState;
-            BoundStatement assignFinishedState =
-                F.ExpressionStatement(F.AssignmentExpression(F.Field(F.This(), stateField), F.Literal(StateMachineState.FinishedState)));
-
-            // builder.SetException(ex);  OR  if (this.combinedTokens != null) this.combinedTokens.Dispose(); _promiseOfValueOrEnd.SetException(ex);
-            BoundStatement callSetException = GenerateSetExceptionCall(exceptionLocal);
-
-            return new BoundCatchBlock(
-                F.Syntax,
-                ImmutableArray.Create(exceptionLocal),
-                F.Local(exceptionLocal),
-                exceptionLocal.Type,
-                exceptionFilterPrologueOpt: null,
-                exceptionFilterOpt: null,
-                body: F.Block(
-                    assignFinishedState, // _state = finishedState;
-                    GenerateHoistedLocalsCleanup(hoistedLocals),
-                    callSetException, // builder.SetException(ex);  OR  _promiseOfValueOrEnd.SetException(ex);
-                    GenerateReturn(false)), // return;
-                isSynthesizedAsyncCatchAll: true);
-        }
-
-        protected BoundStatement GenerateHoistedLocalsCleanup(ImmutableArray<StateMachineFieldSymbol> hoistedLocals)
+        protected virtual BoundStatement GenerateHoistedLocalsCleanupForExit(ImmutableArray<StateMachineFieldSymbol> rootHoistedLocals)
```

Comment:
>GenerateHoistedLocalsCleanupForExit [](http://example.com/codeflow?start=41&length=35)

Consider removing "HoistedLocals" from the name to completely avoid any confusion that the clean up is limited to the locals in the ```rootHoistedLocals``` parameter.
```
        protected virtual BoundStatement GenerateCleanupForExit(ImmutableArray<StateMachineFieldSymbol> rootHoistedLocals)
```